llvm-bazel/llvm-project-overlay/llvm/BUILD

```diff
@@ -30,2213 +14,1351 @@ exports_files(["LICENSE.TXT"])
-        ),
+        "@LLVM_ENUM_ASM_PARSERS@": "\n".join(["LLVM_ASM_PARSER({})".format(t) for t in llvm_target_asm_parsers]),
     },
 )
 
+llvm_target_disassemblers = [
+    "AArch64",
+    "AMDGPU",
+    "ARM",
+    "BPF",
+    "Hexagon",
+    "Lanai",
+    "PowerPC",
+    "RISCV",
+    "Sparc",
+    "WebAssembly",
+    "X86",
+]
+
 template_rule(
     name = "disassemblers_def_gen",
     src = "include/llvm/Config/Disassemblers.def.in",
     out = "include/llvm/Config/Disassemblers.def",
     substitutions = {
-        "@LLVM_ENUM_DISASSEMBLERS@": "\n".join(
-            ["LLVM_DISASSEMBLER({})".format(t) for t in llvm_target_disassemblers],
-        ),
+        "@LLVM_ENUM_DISASSEMBLERS@": "\n".join(["LLVM_DISASSEMBLER({})".format(t) for t in llvm_target_disassemblers]),
     },
 )
 
-# A common library that all LLVM targets depend on.
-# TODO(b/113996071): We need to glob all potentially #included files and stage
-# them here because LLVM's build files are not strict headers clean, and remote
-# build execution requires all inputs to be depended upon.
-cc_library(
-    name = "config",
-    hdrs = glob([
-        "**/*.h",
-        "**/*.def",
-        "**/*.inc.cpp",
-    ]) + [
-        "include/llvm/Config/AsmParsers.def",
-        "include/llvm/Config/AsmPrinters.def",
-        "include/llvm/Config/Disassemblers.def",
-        "include/llvm/Config/Targets.def",
-        "include/llvm/Config/config.h",
-        "include/llvm/Config/llvm-config.h",
-        "include/llvm/Config/abi-breaking.h",
-    ],
-    defines = llvm_defines,
-    includes = ["include"],
-)
-
-# A creator of an empty file include/llvm/Support/VCSRevision.h.
-# This is usually populated by the upstream build infrastructure, but in this
-# case we leave it blank. See upstream revision r300160.
+# TODO: Need to replace this with something that actually extracts the git
+# commit from the LLVM submodule.
 genrule(
-    name = "vcs_revision_gen",
-    srcs = [],
-    outs = ["include/llvm/Support/VCSRevision.h"],
-    cmd = "echo '' > \"$@\"",
+    name = "version_info_gen",
+    outs = ["include/llvm/Config/VersionInfo.h"],
+    cmd = "echo \"#define LLVM_VERSION_INFO \\\"git\\\"\" > $@",
 )
 
-# Rules that apply the LLVM tblgen tool.
-gentbl(
-    name = "attributes_gen",
-    tbl_outs = [("-gen-attrs", "include/llvm/IR/Attributes.inc")],
-    tblgen = ":llvm-tblgen",
-    td_file = "include/llvm/IR/Attributes.td",
-    td_srcs = ["include/llvm/IR/Attributes.td"],
-)
+# These rules substitute fairly generic Linux and x86 flavored configs
+# currently. We could extend it in-place with selects-style switching on
+# platforms, or by extracting parts of the dictionary to be computed in with
+# a separate starlark function where the config switching can be managed.
+#
+# TODO: Priority list of enhancements needed here:
+# - Reflect `opt`, `dbg`, etc. features.
+# - Basic port to Mac for convenience.
+# - Extract into something cleaner based on Starlark.
```

Comment:
Ah good point. And I agree that doing it with a script wrapped in starlark. Looks like gn has one. Maybe we could even use the same one: https://github.com/llvm/llvm-project/blob/master/llvm/utils/gn/build/write_cmake_config.py

That can be a follow-up though :-)

Reply:
Ah I just realized that expand_cmake_vars is a script we already have that does this. It might also help to address the issues I'm hitting with terminfo and such with IREE.

Reply:
Yeah good point. We can do this entirely in starlark and have a nicely formatted substitution actually. Configuring based on selects also sounds good.

Follow-up sounds good for both